Deprecate Filename.chop_suffix in favor of check_and_chop_suffix.

whitequark opened the pull request:
See https://caml.inria.fr/mantis/view.php?id=7812.

~The defined behavior on erroneous input is to always raise Invalid_argument. This is chosen as opposed to e.g. doing nothing because it matches the de-facto behavior on earlier versions (which would be to raise a different exception) and avoids surprises when developing code for 4.08+ and running it on an earlier version.~ See the PR discussion for details.

alainfrisch commented on stdlib/filename.mli:
```diff
@@ -46,8 +46,10 @@ val check_suffix : string -> string -> bool
-   end with the suffix [suff]. *)
+   the filename [name].
+
+   @since 4.08 If [name] does not end with the suffix [suff],
```
If one wants to fully specify the behavior, one should explain what "ending with the suffix" means (also for other existing functions).  This is not straightforward: for Windows ports, the comparison is case-insensitive on ASCII characters; but not, for instance on macOS (whose file-system is also usually case-insensitive, I believe).

whitequark replied:
I've specified `chop_suffix` in terms of `check_suffix`. Your suggestion could be done as a separate patch if there is desire to do so.

alainfrisch commented:
>  because it matches the de-facto behavior on earlier versions (which would be to raise a different exception)

What do you mean?  Earlier versions fails only when the suffix is longer than the filename.

Since chop_suffix is likely to always be used in conjunction with check_suffix, I'm wondering whether it wouldn't be better to deprecate this function and expose:

````ocaml
  val chop_suffix_opt: string -> string -> string option
````

or even, using labels to avoid the confusion between the two string arguments:

````ocaml
  val chop_suffix_opt: suffix:string -> string -> string option
````

whitequark commented:
> Earlier versions fails only when the suffix is longer than the filename.

Yes, that's what I meant--we should not change the behavior when the suffix is longer than the filename. If we changed `chop_suffix` to do nothing when the suffix is not present  (such as when the suffix is longer than the filename), then code wirtten against 4.08 and ran against (e.g.) 4.07 would silently exhibit very surprising behavior.

> Since chop_suffix is likely to always be used in conjunction with check_suffix, I'm wondering whether it wouldn't be better to deprecate this function and expose:

I have no opinion on this. I will implement whichever variant there is consensus for.

damiendoligez commented:
I would expect a name like `check_and_chop_suffix` or `check_and_remove_suffix`.

whitequark committed:
Deprecate Filename.chop_suffix in favor of check_and_chop_suffix. (PR7812)

Filename.chop_suffix is currently documented as having undefined
behavior on inputs that don't end with the suffix. The replacement
does not have this drawback.

whitequark committed:
Fix uses of deprecated Filename.chop_suffix in the compiler.

whitequark commented:
Updated.

alainfrisch commented:
- Typo sufffix -> suffix.

- The name of the function (check_...) suggests it is intended to be used on arbitrary inputs; if the suffix does not match, this is not a programmer error, and the function should thus not raise Invalid_argument.  

- The current direction in stdlib is to prefer functions returning options over raising exceptions, and it is a canonical use of named arguments to avoid confusion between arguments of the same type when there is no natural ordering between them.  So I propose:

````ocaml
val check_and_chop_suffix : suffix:string -> string -> string option
````

alainfrisch reviewed (changes requested):
See previous comments.

nojb commented:
See also https://caml.inria.fr/mantis/view.php?id=7867 for a bug that could have been avoided with the option-returning function.

alainfrisch commented:
Superseded by https://github.com/ocaml/ocaml/pull/2125
